### ISSUE TITLE
`NormalizeData` broken in Seurat V5 on BPCells IterableMatrix

### ISSUE THREAD

**Issue by mjz1** [satijalab/seurat#7423]
On the latest V5 (4.9.9.9048 -- satijalab/seurat@6b8026a).

```Error in LogNormalize(data = object, scale.factor = scale.factor, margin = cmargin, : unused arguments (margin = cmargin, layer = "data")```

Reproducible example below:

``` r
library(Seurat)
#> Loading required package: SeuratObject
#> Loading required package: sp
#> The legacy packages maptools, rgdal, and rgeos, underpinning this package
#> will retire shortly. Please refer to R-spatial evolution reports on
#> https://r-spatial.org/r/2023/05/15/evolution4.html for details.
#> This package is now running under evolution status 0
#> 
#> Attaching package: 'SeuratObject'
#> The following objects are masked from 'package:base':
#> 
#>     intersect, saveRDS
#> Loading Seurat v5 beta version 
#> To maintain compatibility with previous workflows, new Seurat objects will use the previous object structure by default
#> To use new Seurat v5 assays: Please run: options(Seurat.object.assay.version = 'v5')
library(SeuratData)
#> ── Installed datasets ──────────────────────────────── SeuratData v0.2.2.9001 ──
#> ✔ bonemarrowref 1.0.0                   ✔ pbmc3k        3.1.4
#> ✔ lungref       2.0.0                   ✔ pbmcref       1.0.0
#> ────────────────────────────────────── Key ─────────────────────────────────────
#> ✔ Dataset loaded successfully
#> ❯ Dataset built with a newer version of Seurat than installed
#> ❓ Unknown version of Seurat installed
library(BPCells)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
options(Seurat.object.assay.version = "v5")

pbmc3k <- LoadData("pbmc3k")
#> Validating object structure
#> Updating object slots
#> Ensuring keys are in the proper structure
#> Warning: Assay RNA changing from Assay to Assay
#> Ensuring keys are in the proper structure
#> Ensuring feature names don't have underscores or pipes
#> Updating slots in RNA
#> Validating object structure for Assay 'RNA'
#> Object representation is consistent with the most current Seurat version
#> Warning: Assay RNA changing from Assay to Assay5

NormalizeData(pbmc3k)
#> Normalizing layer: counts
#> Warning: The following arguments are not used: layer
#> An object of class Seurat 
#> 13714 features across 2700 samples within 1 assay 
#> Active assay: RNA (13714 features, 0 variable features)
#>  2 layers present: counts, data


write_matrix_dir(
  mat = pbmc3k[["RNA"]]$counts,
  dir = "./tmp"
)
#> Warning: Matrix compression performs poorly with non-integers.
#> • Consider calling convert_matrix_type if a compressed integer matrix is intended.
#> This message is displayed once every 8 hours.
#> 13714 x 2700 IterableMatrix object with class MatrixDir
#> 
#> Row names: AL627309.1, AP006222.2 ... SRSF10.1
#> Col names: AAACATACAACCAC, AAACATTGAGCTAC ... TTTGCATGCCTCAC
#> 
#> Data type: double
#> Storage order: column major
#> 
#> Queued Operations:
#> 1. Load compressed matrix from directory /tmp/RtmpCtXSzT/reprex-c6fe536cee01-grave-cobra/tmp

pbmc3k[["RNA"]]$counts <- open_matrix_dir("./tmp")

pbmc3k[["RNA"]]$counts
#> 13714 x 2700 IterableMatrix object with class RenameDims
#> 
#> Row names: AL627309.1, AP006222.2 ... SRSF10.1
#> Col names: AAACATACAACCAC, AAACATTGAGCTAC ... TTTGCATGCCTCAC
#> 
#> Data type: double
#> Storage order: column major
#> 
#> Queued Operations:
#> 1. Load compressed matrix from directory /tmp/RtmpCtXSzT/reprex-c6fe536cee01-grave-cobra/tmp
#> 2. Select rows: 1, 2 ... 13714 and cols: 1, 2 ... 2700
#> 3. Reset dimnames

NormalizeData(pbmc3k)
#> Normalizing layer: counts
#> Error in LogNormalize(data = object, scale.factor = scale.factor, margin = cmargin, : unused arguments (margin = cmargin, layer = "data")
```

<sup>Created on 2023-06-07 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.3.0 (2023-04-21)
#>  os       Ubuntu 22.04.2 LTS
#>  system   x86_64, linux-gnu
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       America/New_York
#>  date     2023-06-07
#>  pandoc   2.19.2 @ /juno/work/shah/users/zatzmanm/miniconda3/bin/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package                  * version    date (UTC) lib source
#>  abind                      1.4-5      2016-07-21 [1] RSPM (R 4.3.0)
#>  BiocGenerics               0.46.0     2023-04-25 [1] RSPM (R 4.3.0)
#>  bitops                     1.0-7      2021-04-24 [1] RSPM (R 4.3.0)
#>  bonemarrowref.SeuratData * 1.0.0      2023-01-17 [1] local
#>  BPCells                  * 0.1.0      2023-06-04 [1] Github (bnprks/BPCells@9d1a4ca)
#>  cli                        3.6.1      2023-03-23 [1] RSPM (R 4.3.0)
#>  cluster                    2.1.4      2022-08-22 [3] CRAN (R 4.3.0)
#>  codetools                  0.2-19     2023-02-01 [3] CRAN (R 4.3.0)
#>  colorspace                 2.1-0      2023-01-23 [1] RSPM (R 4.3.0)
#>  cowplot                    1.1.1      2020-12-30 [1] RSPM (R 4.3.0)
#>  crayon                     1.5.2      2022-09-29 [1] RSPM (R 4.3.0)
#>  data.table                 1.14.8     2023-02-17 [1] RSPM (R 4.3.0)
#>  deldir                     1.0-9      2023-05-17 [1] RSPM (R 4.3.0)
#>  digest                     0.6.31     2022-12-11 [1] RSPM (R 4.3.0)
#>  dotCall64                  1.0-2      2022-10-03 [1] RSPM (R 4.3.0)
#>  dplyr                    * 1.1.2      2023-04-20 [1] RSPM (R 4.3.0)
#>  ellipsis                   0.3.2      2021-04-29 [1] RSPM (R 4.3.0)
#>  evaluate                   0.21       2023-05-05 [1] RSPM (R 4.3.0)
#>  fansi                      1.0.4      2023-01-22 [1] RSPM (R 4.3.0)
#>  fastDummies                1.6.3      2020-11-29 [1] RSPM (R 4.3.0)
#>  fastmap                    1.1.1      2023-02-24 [1] RSPM (R 4.3.0)
#>  fitdistrplus               1.1-11     2023-04-25 [1] RSPM (R 4.3.0)
#>  fs                         1.6.2      2023-04-25 [1] RSPM (R 4.3.0)
#>  future                     1.32.0     2023-03-07 [1] RSPM (R 4.3.0)
#>  future.apply               1.11.0     2023-05-21 [1] RSPM (R 4.3.0)
#>  generics                   0.1.3      2022-07-05 [1] RSPM (R 4.3.0)
#>  GenomeInfoDb               1.36.0     2023-04-25 [1] RSPM (R 4.3.0)
#>  GenomeInfoDbData           1.2.10     2023-04-30 [1] RSPM (R 4.3.0)
#>  GenomicRanges              1.52.0     2023-04-25 [1] RSPM (R 4.3.0)
#>  ggplot2                    3.4.2      2023-04-03 [1] RSPM (R 4.3.0)
#>  ggrepel                    0.9.3      2023-02-03 [1] RSPM (R 4.3.0)
#>  ggridges                   0.5.4      2022-09-26 [1] RSPM (R 4.3.0)
#>  globals                    0.16.2     2022-11-21 [1] RSPM (R 4.3.0)
#>  glue                       1.6.2      2022-02-24 [1] RSPM (R 4.3.0)
#>  goftest                    1.2-3      2021-10-07 [1] RSPM (R 4.3.0)
#>  gridExtra                  2.3        2017-09-09 [1] RSPM (R 4.3.0)
#>  gtable                     0.3.3      2023-03-21 [1] RSPM (R 4.3.0)
#>  htmltools                  0.5.5      2023-03-23 [1] RSPM (R 4.3.0)
#>  htmlwidgets                1.6.2      2023-03-17 [1] RSPM (R 4.3.0)
#>  httpuv                     1.6.11     2023-05-11 [1] RSPM (R 4.3.0)
#>  httr                       1.4.6      2023-05-08 [1] RSPM (R 4.3.0)
#>  ica                        1.0-3      2022-07-08 [1] RSPM (R 4.3.0)
#>  igraph                     1.4.3      2023-05-22 [1] RSPM (R 4.3.0)
#>  IRanges                    2.34.0     2023-04-25 [1] RSPM (R 4.3.0)
#>  irlba                      2.3.5.1    2022-10-03 [1] RSPM (R 4.3.0)
#>  jsonlite                   1.8.5      2023-06-05 [1] RSPM (R 4.3.0)
#>  KernSmooth                 2.23-20    2021-05-03 [3] CRAN (R 4.3.0)
#>  knitr                      1.43       2023-05-25 [1] RSPM (R 4.3.0)
#>  later                      1.3.1      2023-05-02 [1] RSPM (R 4.3.0)
#>  lattice                    0.21-8     2023-04-05 [3] CRAN (R 4.3.0)
#>  lazyeval                   0.2.2      2019-03-15 [1] RSPM (R 4.3.0)
#>  leiden                     0.4.3      2022-09-10 [1] RSPM (R 4.3.0)
#>  lifecycle                  1.0.3      2022-10-07 [1] RSPM (R 4.3.0)
#>  listenv                    0.9.0      2022-12-16 [1] RSPM (R 4.3.0)
#>  lmtest                     0.9-40     2022-03-21 [1] RSPM (R 4.3.0)
#>  lungref.SeuratData       * 2.0.0      2023-01-20 [1] local
#>  magrittr                   2.0.3      2022-03-30 [1] RSPM (R 4.3.0)
#>  MASS                       7.3-59     2023-04-21 [3] RSPM (R 4.3.0)
#>  Matrix                     1.5-4.1    2023-05-18 [1] RSPM (R 4.3.0)
#>  matrixStats                1.0.0      2023-06-02 [1] RSPM (R 4.3.0)
#>  mime                       0.12       2021-09-28 [1] RSPM (R 4.3.0)
#>  miniUI                     0.1.1.1    2018-05-18 [1] RSPM (R 4.3.0)
#>  munsell                    0.5.0      2018-06-12 [1] RSPM (R 4.3.0)
#>  nlme                       3.1-162    2023-01-31 [3] CRAN (R 4.3.0)
#>  parallelly                 1.36.0     2023-05-26 [1] RSPM (R 4.3.0)
#>  patchwork                  1.1.2      2022-08-19 [1] RSPM (R 4.3.0)
#>  pbapply                    1.7-0      2023-01-13 [1] RSPM (R 4.3.0)
#>  pbmc3k.SeuratData        * 3.1.4      2023-06-07 [1] local
#>  pbmcref.SeuratData       * 1.0.0      2023-01-17 [1] local
#>  pillar                     1.9.0      2023-03-22 [1] RSPM (R 4.3.0)
#>  pkgconfig                  2.0.3      2019-09-22 [1] RSPM (R 4.3.0)
#>  plotly                     4.10.2     2023-06-03 [1] RSPM (R 4.3.0)
#>  plyr                       1.8.8      2022-11-11 [1] RSPM (R 4.3.0)
#>  png                        0.1-8      2022-11-29 [1] RSPM (R 4.3.0)
#>  polyclip                   1.10-4     2022-10-20 [1] RSPM (R 4.3.0)
#>  progressr                  0.13.0     2023-01-10 [1] RSPM (R 4.3.0)
#>  promises                   1.2.0.1    2021-02-11 [1] RSPM (R 4.3.0)
#>  purrr                      1.0.1      2023-01-10 [1] RSPM (R 4.3.0)
#>  R.cache                    0.16.0     2022-07-21 [1] RSPM (R 4.3.0)
#>  R.methodsS3                1.8.2      2022-06-13 [1] RSPM (R 4.3.0)
#>  R.oo                       1.25.0     2022-06-12 [1] RSPM (R 4.3.0)
#>  R.utils                    2.12.2     2022-11-11 [1] RSPM (R 4.3.0)
#>  R6                         2.5.1      2021-08-19 [1] RSPM (R 4.3.0)
#>  RANN                       2.6.1      2019-01-08 [1] RSPM (R 4.3.0)
#>  rappdirs                   0.3.3      2021-01-31 [1] RSPM (R 4.3.0)
#>  RColorBrewer               1.1-3      2022-04-03 [1] RSPM (R 4.3.0)
#>  Rcpp                       1.0.10     2023-01-22 [1] RSPM (R 4.3.0)
#>  RcppAnnoy                  0.0.20     2022-10-27 [1] RSPM (R 4.3.0)
#>  RcppHNSW                   0.4.1      2022-07-18 [1] RSPM (R 4.3.0)
#>  RCurl                      1.98-1.12  2023-03-27 [1] RSPM (R 4.3.0)
#>  reprex                     2.0.2      2022-08-17 [1] RSPM (R 4.3.0)
#>  reshape2                   1.4.4      2020-04-09 [1] RSPM (R 4.3.0)
#>  reticulate                 1.29       2023-06-05 [1] RSPM (R 4.3.0)
#>  rlang                      1.1.1      2023-04-28 [1] RSPM (R 4.3.0)
#>  rmarkdown                  2.22       2023-06-01 [1] RSPM (R 4.3.0)
#>  ROCR                       1.0-11     2020-05-02 [1] RSPM (R 4.3.0)
#>  RSpectra                   0.16-1     2022-04-24 [1] RSPM (R 4.3.0)
#>  rstudioapi                 0.14       2022-08-22 [1] RSPM (R 4.3.0)
#>  Rtsne                      0.16       2022-04-17 [1] RSPM (R 4.3.0)
#>  S4Vectors                  0.38.1     2023-05-02 [1] RSPM (R 4.3.0)
#>  scales                     1.2.1      2022-08-20 [1] RSPM (R 4.3.0)
#>  scattermore                1.1        2023-05-17 [1] RSPM (R 4.3.0)
#>  sctransform                0.3.5      2022-09-21 [1] RSPM (R 4.3.0)
#>  sessioninfo                1.2.2      2021-12-06 [1] RSPM (R 4.3.0)
#>  Seurat                   * 4.9.9.9048 2023-06-07 [1] Github (satijalab/seurat@6b8026a)
#>  SeuratData               * 0.2.2.9001 2023-04-07 [1] Github (satijalab/seurat-data@dd7d0df)
#>  SeuratObject             * 4.9.9.9084 2023-05-08 [1] Github (mojaveazure/seurat-object@185884a)
#>  shiny                      1.7.4      2022-12-15 [1] RSPM (R 4.3.0)
#>  sp                       * 1.6-1      2023-05-31 [1] RSPM (R 4.3.0)
#>  spam                       2.9-1      2022-08-07 [1] RSPM (R 4.3.0)
#>  spatstat.data              3.0-1      2023-03-12 [1] RSPM (R 4.3.0)
#>  spatstat.explore           3.2-1      2023-05-13 [1] RSPM (R 4.3.0)
#>  spatstat.geom              3.2-1      2023-05-09 [1] RSPM (R 4.3.0)
#>  spatstat.random            3.1-5      2023-05-11 [1] RSPM (R 4.3.0)
#>  spatstat.sparse            3.0-1      2023-03-12 [1] RSPM (R 4.3.0)
#>  spatstat.utils             3.0-3      2023-05-09 [1] RSPM (R 4.3.0)
#>  stringi                    1.7.12     2023-01-11 [1] RSPM (R 4.3.0)
#>  stringr                    1.5.0      2022-12-02 [1] RSPM (R 4.3.0)
#>  styler                     1.10.1     2023-06-05 [1] RSPM (R 4.3.0)
#>  survival                   3.5-5      2023-03-12 [1] RSPM (R 4.3.0)
#>  tensor                     1.5        2012-05-05 [1] RSPM (R 4.3.0)
#>  tibble                     3.2.1      2023-03-20 [1] RSPM (R 4.3.0)
#>  tidyr                      1.3.0      2023-01-24 [1] RSPM (R 4.3.0)
#>  tidyselect                 1.2.0      2022-10-10 [1] RSPM (R 4.3.0)
#>  utf8                       1.2.3      2023-01-31 [1] RSPM (R 4.3.0)
#>  uwot                       0.1.14     2022-08-22 [1] RSPM (R 4.3.0)
#>  vctrs                      0.6.2      2023-04-19 [1] RSPM (R 4.3.0)
#>  viridisLite                0.4.2      2023-05-02 [1] RSPM (R 4.3.0)
#>  withr                      2.5.0      2022-03-03 [1] RSPM (R 4.3.0)
#>  xfun                       0.39       2023-04-20 [1] RSPM (R 4.3.0)
#>  xtable                     1.8-4      2019-04-21 [1] RSPM (R 4.3.0)
#>  XVector                    0.40.0     2023-04-25 [1] RSPM (R 4.3.0)
#>  yaml                       2.3.7      2023-01-23 [1] RSPM (R 4.3.0)
#>  zlibbioc                   1.46.0     2023-04-25 [1] RSPM (R 4.3.0)
#>  zoo                        1.8-12     2023-04-13 [1] RSPM (R 4.3.0)
#> 
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

**Comment by mjz1** [comment]
This workaround seems to work in the meantime:

```r 
pbmc3k[["RNA"]]$data <- Seurat:::LogNormalize.IterableMatrix(pbmc3k[["RNA"]]$counts)
```

**Comment by Gesmira** [comment]
Hi, 
Thanks for reporting this! The error should now be fixed in the seurat5 branch, Version: 4.9.9.9049. Please restart your session and reinstall. 

```
remotes::install_github("satijalab/seurat", "seurat5", quiet = TRUE)
```
